kbkdf/src/lib.rs

```diff
@@ -0,0 +1,384 @@
+    // length of the bit string is specified by the encoding method for the input data.
+    type L;
+
+    // h - An integer that indicates the length (in bits) of the output of a single invocation of the
+    // PRF.
+    type H;
+}
+
+impl<OutputLen, PrfOutputLen> KbkdfUser for KbkdfCore<OutputLen, PrfOutputLen>
+where
+    OutputLen: ArrayLength<u8> + Mul<U8>,
+    <OutputLen as Mul<U8>>::Output: Unsigned,
+    PrfOutputLen: ArrayLength<u8> + Mul<U8>,
+    <PrfOutputLen as Mul<U8>>::Output: Unsigned,
+{
+    type L = op!(OutputLen * U8);
+    type H = op!(PrfOutputLen * U8);
+}
+
+/// [`Kbkdf`] is a trait representing a mode of KBKDF.
+/// It takes multiple arguments:
+///  - Prf - the Pseudorandom Function to derive keys from
+///  - K - the expected output length of the newly derived key
+///  - R - An integer (1 <= r <= 32) that indicates the length of the binary encoding of the counter i
+///        as an integer in the interval [1, 2r − 1].
+pub trait Kbkdf<Prf, K, R: sealed::R = U32>
+where
+    Prf: Mac + KeyInit,
+    K: KeySizeUser,
+    K::KeySize: ArrayLength<u8> + Mul<U8>,
+    <K::KeySize as Mul<U8>>::Output: Unsigned,
+    Prf::OutputSize: ArrayLength<u8> + Mul<U8>,
+    <Prf::OutputSize as Mul<U8>>::Output: Unsigned,
+{
+    /// Derives `key` from `kin` and other parameters.
+    fn derive(
+        &self,
+        kin: &GenericArray<u8, Prf::KeySize>,
+        use_l: bool,
+        use_separator: bool,
+        label: &[u8],
+        context: &[u8],
+    ) -> Result<GenericArray<u8, K::KeySize>, Error> {
+        // n - An integer whose value is the number of iterations of the PRF needed to generate L
+        // bits of keying material
+        let n: u32 = Wrapping(<KbkdfCore<K::KeySize, Prf::OutputSize> as KbkdfUser>::L::U32)
+            .div_ceil(Wrapping(
+                <KbkdfCore<K::KeySize, Prf::OutputSize> as KbkdfUser>::H::U32,
+            ))
+            .0;
+
+        if n as usize > 2usize.pow(R::U32) - 1 {
+            return Err(Error::InvalidRequestSize);
+        }
+
+        let mut output = GenericArray::<u8, K::KeySize>::default();
+        let mut builder = output.as_mut_slice();
+
+        let mut ki = None;
+        self.input_iv(&mut ki);
+
+        for counter in 1..=n {
+            let mut h = <Prf as Mac>::new(kin);
+
+            if Self::FEEDBACK_KI {
+                if let Some(ki) = ki {
+                    h.update(ki.as_slice());
+                }
+            }
+
+            // counter encoded as big endian u32
+            // Type parameter R encodes how large the value is to be (either U8, U16, U24, or U32)
+            //
+            // counter = 1u32 ([0, 0, 0, 1])
+            //                     \-------/
+            //                      R = u24
+            h.update(&counter.to_be_bytes()[(4 - R::USIZE / 8)..]);
+
+            // Fixed input data
+            h.update(label);
+            if use_separator {
```

Comment:
This behavior is not standard, but TPM specification relies on KBKDF and calls for separator to be optional (https://trustedcomputinggroup.org/wp-content/uploads/TCG_TPM2_r1p59_Part1_Architecture_pub.pdf#page=73).
```
Added only if Label is not present or if the last octet of Label is not zero.
```

For a time I intended to provide that behavior as a type parameter, but I'm not sure now. I'd love opinion.